From ada404dfd25e63fc4522ac548932c21cab009dbb Mon Sep 17 00:00:00 2001
From: Johannes Lorenz <j.git@lorenz-ho.me>
Date: Tue, 15 Dec 2020 20:51:06 +0100
Subject: [PATCH 1680/1936] Remove Controller::operator=

`Controller` contains a member reference, so the default constructor
would not work. The previous implementation used `memcpy` as a
workaround for that in `Controller::operator=`. This triggered a
compiler warning that `memcpy` of an object with non-trivial
copy-assignment should be avoided.

This commit removes this `Controller::operator=` workaround, and clones
the `Controller` object by destroying it explicitly and then using the
copy ctor via placement new.

Furthermore, remove `Controller::~Controller` (Rule of zero).
---
 src/Misc/Part.cpp         |  4 +++-
 src/Params/Controller.cpp | 10 ----------
 src/Params/Controller.h   |  3 +--
 3 files changed, 4 insertions(+), 13 deletions(-)

diff --git a/src/Misc/Part.cpp b/src/Misc/Part.cpp
index 356dbcc5..062204c8 100644
--- a/src/Misc/Part.cpp
+++ b/src/Misc/Part.cpp
@@ -353,7 +353,9 @@ void Part::cloneTraits(Part &p) const
     CLONE(Plegatomode);
     CLONE(Pkeylimit);
 
-    CLONE(ctl);
+    // Controller has a refence, so it can not be re-assigned
+    // So, destroy and reconstruct it.
+    p.ctl.~Controller(); new (&p.ctl) Controller(this->ctl);
 }
 
 void Part::defaults()
diff --git a/src/Params/Controller.cpp b/src/Params/Controller.cpp
index dae8737d..a61bf41d 100644
--- a/src/Params/Controller.cpp
+++ b/src/Params/Controller.cpp
@@ -93,16 +93,6 @@ Controller::Controller(const SYNTH_T &synth_, const AbsTime *time_)
     resetall();
 }
 
-Controller &Controller::operator=(const Controller &c)
-{
-    //just pretend that undefined behavior doesn't exist...
-    memcpy(this, &c, sizeof(Controller));
-    return *this;
-}
-
-Controller::~Controller()
-{}
-
 void Controller::defaults()
 {
     pitchwheel.bendrange = 200; //2 halftones
diff --git a/src/Params/Controller.h b/src/Params/Controller.h
index e95c9084..b10e0497 100644
--- a/src/Params/Controller.h
+++ b/src/Params/Controller.h
@@ -25,8 +25,7 @@ class Controller
 {
     public:
         Controller(const SYNTH_T &synth, const AbsTime *time = nullptr);
-        Controller&operator=(const Controller &c);
-        ~Controller();
+
         void resetall();
 
         void add2XML(XMLwrapper& xml);
-- 
2.47.0

